httpreplay/cobweb.py

```diff
@@ -175,3 +180,245 @@ def handle(self, s, ts, protocol, sent, recv):
+
+        if len(data) < 2:
+            return
+
+        arg_first = data[1].lower()
+        if arg_first not in auth_handlers:
+            return
+
+        self.request.auth_type = arg_first
+
+        if len(data) > 2:
+            arg_second = data[2]
+            handler = auth_handlers[arg_first]
+            handler(arg_second)
+
+    def handle_auth_plain(self, arg):
+        try:
+            user_pass = filter(None, arg.decode("base64").split("\x00"))
+            if len(user_pass) < 2:
+                return
+
+            self.request.username = user_pass[0]
+            self.request.password = user_pass[1]
+        except binascii.Error:
+            return
+
+    def handle_auth_login(self, arg):
+        try:
+            self.request.username = arg.decode("base64")
+        except binascii.Error:
+            return
+
+    def handle_auth_cram_md5(self, arg):
+        user_challange = arg.decode("base64").split(" ", 1)
+        if len(user_challange) < 2:
+            return
+
+        self.request.username = user_challange[0]
+
+    def handle_auth_serv_response(self, data):
+        """
+        If not all credentials were passed as an argument to an auth
+        command, they will be later sent when the server requests them. If
+        that happens, this function extracts them
+        """
+        if len(data) < 1:
+            return
+
+        client_response = data[0]
+
+        if self.request.auth_type == "plain":
+            self.handle_auth_plain(client_response)
+
+        elif self.request.auth_type == "cram-md5":
+            self.handle_auth_cram_md5(client_response)
+
+        elif self.request.auth_type == "login":
+            if "UGFzc3dvcmQ6" in self.message:
+                try:
+                    self.request.password = client_response.decode("base64")
+                except binascii.Error:
+                    return
+            elif "VXNlcm5hbWU6" in self.message:
+                try:
+                    self.request.username = client_response.decode("base64")
+                except binascii.Error:
+                    return
+
+    def parse_request(self, request):
+        """"
+        Parses the requests sent by an smtp client. Values are stored in
+        a SmtpRequest object.
+        """
+        self.request.raw.append(request)
+
+        if self.command != "data":
+            data = request.split(None)
+        else:
+            data = request
+
+        if len(data) < 1:
+            return
+
+        cmd = data[0].lower()
+
+        # If no valid command is found, see if there are
+        # any actions to be performed for the last received response code
+        if cmd not in self._commands:
+            if self.rescode in self._res_codes:
+                handle = self._res_codes[self.rescode]
+                handle(data)
+            return
+
+        else:
+            self.command = cmd
+            handler = self._commands[cmd]
+            if handler is not None:
```

Comment:
[part 2 of my comment] ... and simply replace this with `if cmd in self._commands:` then you'll have the same functionality without having to specify each known command :-)